### PR TITLE
Wire up <UseTools /> to use OpenAI function calls if available

### PR DESCRIPTION
**Pull request by hessamb**
This wires OpenAI function calls through to <UseTools /> to rely on them for function calling and consuming their responses.

Thanks to @farzadab for walking through the hurdles of typescript on this with me!

### CODE REVIEW

**Review by farzadab — APPROVED** [review]
LGTM

File: packages/ai-jsx/src/batteries/use-tools.tsx [hunk 1]
```diff
@@ -183,6 +190,85 @@ export interface UseToolsProps {
...
+  messages.push(<UserMessage>{props.query}</UserMessage>);
+
+  do {
+    const model_response = (
+      <ChatCompletion
+        functionDefinitions={Object.entries(props.tools).map(([toolName, tool]) => ({
+          name: toolName,
+          description: tool.description,
+          parameters: tool.parameters,
+        }))}
```
Comment (NickHeiner) [inline packages/ai-jsx/src/batteries/use-tools.tsx:227]: Should we just map the type of `props.tools` precisely the same as `functionDefinitions`?

File: packages/ai-jsx/src/batteries/use-tools.tsx [hunk 2]
```diff
@@ -183,6 +190,85 @@ export interface UseToolsProps {
...
+    if (e.code === ErrorCode.ChatModelDoesntSupportFunctions) {
+      return <UseToolsPromptEngineered {...props} />;
+    }
+    throw e;
+  }
+}
+
+/**
+ * An implementation of <UseTools/> that uses {@link ChatCompletion}'s {@link FunctionDefinition} capability to call
+ * functions. The chat model in scope must support Function Calls for this component.
```
Comment (NickHeiner) [inline packages/ai-jsx/src/batteries/use-tools.tsx:207]: Is this a user-facing API? If so, let's link out to the tutorials page / docs content for using tools. If not, let's mark it as `@hidden`.

File: packages/ai-jsx/src/batteries/use-tools.tsx [hunk 3]
```diff
@@ -183,6 +190,85 @@ export interface UseToolsProps {
...
+  ];
+  if (props.fallback) {
+    messages.push(
+      <SystemMessage>Here's the fallback strategy/message if something failed: {props.fallback}</SystemMessage>
+    );
+  }
+  messages.push(<UserMessage>{props.query}</UserMessage>);
+
+  do {
+    const model_response = (
```
Comment (NickHeiner) [inline packages/ai-jsx/src/batteries/use-tools.tsx:221]:
```suggestion
    const modelResponse = (
```

Please don't make me turn on the lint rule enforcing camelCase. :smile:

File: packages/ai-jsx/src/batteries/use-tools.tsx [hunk 4]
```diff
@@ -183,6 +190,85 @@ export interface UseToolsProps {
...
+        {messages}
+      </ChatCompletion>
+    );
+
+    const renderResult = yield* render(model_response, { stop: (el) => el.tag == FunctionCall });
+
+    if (isElement(renderResult[0])) {
+      // Model has generated a <FunctionCall/> element.
+      if (renderResult.length > 1 || renderResult[0].tag != FunctionCall) {
+        throw new AIJSXError(
```
Comment (NickHeiner) [inline packages/ai-jsx/src/batteries/use-tools.tsx:238]: You can also use the `metadata` argument to the `AIJSXError` constructor to pass the `renderResult` directly.

File: packages/ai-jsx/src/core/errors.ts [hunk 5]
```diff
@@ -20,6 +20,7 @@ export enum ErrorCode {
   NestedAIUIStreamsAreNotSupported = 1015,
   UnknownUIComponentId = 1016,
   UnknownSerializedComponentType = 1017,
+  ChatModelDoesntSupportFunctions = 1018,
```
Comment (NickHeiner) [inline packages/ai-jsx/src/core/errors.ts:23]:
```suggestion
  ChatModelDoesNotSupportFunctions = 1018,
```

**Review by NickHeiner — APPROVED** [review]
I did a pretty quick skim of this but it seems good!

**Comment by NickHeiner** [comment]
Do we need to update the docs/tutorials in any way to reflect this?

File: packages/ai-jsx/src/batteries/use-tools.tsx (hunk 4 above)
Comment (hessamb) [reply packages/ai-jsx/src/batteries/use-tools.tsx:238]: Looks like `metadata` needs to be Jsonifable; whereas `renderResult` is not.

File: packages/ai-jsx/src/batteries/use-tools.tsx (hunk 1 above)
Comment (hessamb) [reply packages/ai-jsx/src/batteries/use-tools.tsx:224]: Good callout. I changed functionDefinitions to match with props.tools

**Pull request merged**